Fixes

matthijskooijman opened the pull request:
This PR fixes a number of issues I was having while trying to run my code inside arduino_ci. Fixes are various, see the git commit messages for details.

Note that I originally was not running arduino_ci itself, but instead using arduino_ci's `cpp/arduino` code separately to allow running my firmware on a regular PC in a sort of "simulator" mode, but all of the issues I encountered were easy to reproduce in a unit test as well.

Here's a small "library" that reproduces most of the problems solved in this PR (it's really just an empty library with only unittests, but if you run `arduino_ci_remote.rb --skip-compilation` inside it, you'll get the errors): 
[testlib.zip](https://github.com/Arduino-CI/arduino_ci/files/5129501/testlib.zip). Full output is here: [errors.txt](https://github.com/Arduino-CI/arduino_ci/files/5129505/errors.txt)


Some of the tests in this library might be useful to include in the `TestSomething` sample project (which I think is basically the unittests for arudino_ci itself, right?), but I wasn't quite sure which and how (and also, I'm already quite sidetracked off my original project by this), so I left that for now.

Also, I did not add anything to the changelog yet, since that seems like a recipe for merge conflicts. How should this be done? Just add notes to the changelog as part of the commit that makes the change and hope things are merged quickly before conflicts arise? Or do you generate a changelog periodically or at release or so?

## Issues Fixed

This fixes #140, #133 and part of #89.

matthijskooijman commented:
Hm, the CI check seems to fail with two problems:
 - In `pgmspace.h` some functions are not available, like `memccpy`.
 - Some non-nill ruby assertions that I'm note quite sure what to make of

Regarding the missing functions, it seems these are the just a bit less-than-standard functions from string.h, which is correctly included and the more standard functions like `memcpy` *are* available. Looking on my own system, I see that `/usr/lib/string.h` defines `memccpy` conditionally:

```
#if defined __USE_MISC || defined __USE_XOPEN || __GLIBC_USE (ISOC2X)
extern void *memccpy (void *__restrict __dest, const void *__restrict __src,
                      int __c, size_t __n)
     __THROW __nonnull ((1, 2));
#endif /* Misc || X/Open.  */
```

However, looking at my system, unless there are some specific macros defined, at least `__USE_MISC` should be defined by `/usr/include/features.h` (included through `/usr/include/x86_64-linux-gnu/bits/libc-header-start.h`). So maybe there is something weird going on on the testing system? Maybe some Windows-specific thing (same header file, different casing or something)?

I can't reproduce either problem here locally, in any case...

matthijskooijman commented:
I just noticed this PR also fixes #133, so I added that in the commit message and PR description.

While I was pushing, I pushed an extra commit to generate a bit extra output from the failing test.

Looking more closely at the output, I think that the ruby errors are actually just test failures resulting from the compilation failure (it expects a runnable executable, but that was not produced), so there's really just one problem here.

matthijskooijman commented:
@ianfixes, I'm not quite sure what the problem is with the CI build here. Do you happen to have a local cygwin installation you work with where you could investigate locally maybe?

ianfixes commented:
Aah, I was waiting for this to pass the tests before reviewing.  I think [the link to the appveyor build](https://ci.appveyor.com/project/ianfixes/arduino-ci/builds/34866250) got broken when I moved this repo to the "Arduino CI" github org.  Can you try to restart this build by force-pushing the branch?  Or push a whitespace change?

ianfixes commented on cpp/arduino/avr/pgmspace.h:
```diff
@@ -50,46 +91,67 @@ out.each { |l| puts d(l) }
-#define strstr_P(...) ::strstr(__VA_ARGS__)
-#define strlcpy_P(...) ::strlcpy(__VA_ARGS__)
-#define memcmp_P(...) ::memcmp(__VA_ARGS__)
+inline const void * memchr_P(const void *s, int val, size_t len) { return memchr(s, val, len); }
```
What's the advantage to the inline wrapper vs the macro?

matthijskooijman replied:
That it is only interpreted as a top level function, and works better with namespaces. I.e. consider:

```
void *foo = ::memchr_P(s, val, len);
```
Which is top namespace qualified to prevent conflicts, which would expand to `::::memchr(...)` with the macro and break.

Also, functions are better than macros when methods of the same name are used. Consider:
```
class Foo {
  void * memchr_P(const void *s, int val, size_t len);
};
```

The macro would be replaced here and break compilation.

ianfixes replied:
I retract my question... for some reason I thought that `memcpy_P` was defined in Arduino itself as a macro, and that we were going into new territory by making it an actual function.  Just did a quick google search, and it looks like it was a function to begin with -- so I think this is fine.

ianfixes commented on spec/cpp_library_spec.rb:
```diff
@@ -55,6 +55,18 @@ def get_relative_dir(sampleprojects_tests_dir)
   context "test" do
     config = ArduinoCI::CIConfig.default
 
+    puts "\n\nstring.h\n\n"
+    puts File.read("c:/cygwin/usr/include/string.h")
+
+    puts "\n\n_ansi.h\n\n"
+    puts File.read("c:/cygwin/usr/include/_ansi.h")
+
+    puts "\n\nsys/config.h.h\n\n"
+    puts File.read("c:/cygwin/usr/include/sys/config.h")
+
+    puts "\n\nsys/features.h.h\n\n"
+    puts File.read("c:/cygwin/usr/include/sys/features.h")
+
```
I assume this is for debugging and not part of the final PR

matthijskooijman replied:
Yup, everything in the last commit was for debugging.

ianfixes commented:
You should absolutely add lines to the top section of the changelog -- they tend to survive rebases pretty easily and the occasional merge conflict in this file is easy to resolve (pick the existing stuff, then the local stuff...)

If I'm understanding the motivation behind this PR, it's that you were attempting to define your own `min` function (or similar) for a particular data type, but the definition conflicted with the macro?

ianfixes commented:
In general I am happy with all of this if it brings us into closer alignment with the way that these functions behave (compilation-wise and runtime-wise) in official Arduino.

But this fell off my radar because the CI wasn't passing -- I assumed it was still a work in progress.  Is there anything I can offer to help get this to pass its CI checks?

matthijskooijman commented:
W00ps, didn't see all of your comments, not sure what happened there. Sorry for my late reply.

> If I'm understanding the motivation behind this PR, it's that you were attempting to define your own min function (or similar) for a particular data type, but the definition conflicted with the macro?

Yup. In particular, I was including, IIRC `limits.h` (or maybe the boost version of that) which defines a `min()` method, which breaks.

> But this fell off my radar because the CI wasn't passing -- I assumed it was still a work in progress. Is there anything I can offer to help get this to pass its CI checks?

If you use cygwin yourself, maybe try locally to see why compilation fails. There is some weird interaction between system headers IIRC that I couldn't figure out here.

> Can you try to restart this build by force-pushing the branch? Or push a whitespace change?

Done.

matthijskooijman commented:
(Also, feel free to merge #194 separately, I can rebase this PR on top of that after it's merged)

ianfixes commented:
My only access to a cygwin environment is via Appveyor, but multi-platform testing is exactly why I'm using it.  I'm going to try to resolve that merge conflict for you.

ianfixes commented:
Looks like that blew up, but hopefully the output makes it sensible for you to work on

matthijskooijman commented:
Well, it doesn't, unfortunately. It spews a bunch of errors like these:

```
C:/projects/arduino-ci/cpp/arduino/avr/pgmspace.h: In function 'void* memccpy_P3(void*, const void*, int, size_t)':
C:/projects/arduino-ci/cpp/arduino/avr/pgmspace.h:51:84: error: 'memccpy' was not declared in this scope
 inline void *memccpy_P3(void *dest, const void *src, int val, size_t len) { return memccpy(dest, src, val, len); }
```

Indicating some functions are not available, but I can't figure out why not, it looks like they should be available (see also https://github.com/Arduino-CI/arduino_ci/pull/144#issuecomment-680834428).

This did not look like an issue before, because the code used macros, so i.e. `memccpy()` would only be referenced (and break) only if the sketch or library actually used `memmcpy_P()`, but now these are functions, the replacement functions must be available always, so this breaks.

matthijskooijman commented:
I ended up just sidestepping the issue for now: I've removed the offending functions, since they're not quite common and unlikely to actually be used in practice. This should prevent this issue from blocking the other changes, would be a pity if they became stale and out of date. At a later stage (maybe when I have some more time), we can again revisit this.

I've also rebased on top of current master, which worked without conflicts.

Note that I have not tested this at all anymore (I really need to get back to some other work), but the testsuite in appveyor seems to complete succesfully now, so it probably works. You might want to do some testing before merging this, though.

ianfixes commented on misc/default.yml:
```diff
@@ -75,6 +95,10 @@ platforms:
     gcc:
       features:
       defines:
+        - __AVR__
```
I moved a definition of `__AVR__` into the ruby code so it's applied to all boards... was that a mistake on my part?

matthijskooijman replied:
I think so, __AVR__ is normally defined by avr-gcc, so only for the AVR-based Arduino boards. To emulate other boards (e.g. SAMD / Arduino Zero), I think __AVR__ should not be defined, otherwise libraries etc. will not be genuinely emulated and if they test for __AVR__ they will erronously assume that all kind of AVR-specific stuff is available.

ianfixes replied:
aah. Opened https://github.com/Arduino-CI/arduino_ci/issues/202 to track this.

ianfixes commented:
I'll try to resolve the merge conflicts from #194 

In general, my testing is 100% accomplished by the CI machines... so given that your work is passing, I'm inclined to accept it. I'll try to get to that this weekend.

ianfixes commented:
I'm not sure if this is the place to say this, but I'm very grateful for your expertise, your time, and your enthusiasm in contributing to this project.  In the past 2 months this project has sailed past a bunch of technical hurdles that seemed impossible to me earlier this year.

matthijskooijman commented:
> I'll try to resolve the merge conflicts from #194

I just rebased on top of master, dropping the one commit about math macros that was essentially just copied to #194 (with just some changes in whitespace and minor other changes), so I think this should be good again.

Note that I didn't touch the changelog yet. Not 100% sure how you want it, and also no time right now to dive into this, so I'll leave this for you if you don't mind.

> I'm not sure if this is the place to say this, but I'm very grateful for your expertise, your time, and your enthusiasm in contributing to this project. In the past 2 months this project has sailed past a bunch of technical hurdles that seemed impossible to me earlier this year.

Thanks, appreciated. Though I have the idea I haven't been contributing much after an initial splurge of ideas (also since I'm not actually using arduino_ci right now for my own projects yet), but good to hear that my efforts make a difference :-)

ianfixes commented on cpp/arduino/avr/pgmspace.h:
```diff
@@ -50,46 +57,69 @@ out.each { |l| puts d(l) }
+inline int memcmp_PF(const void *s1, uint_farptr_t s2, size_t len) { return memcmp(s1, (const char*)s2, len); }
+inline size_t strlen_P(const char *src) { return strlen(src); }
+
+// TODO: These functions cannot be found on the CYGWIN test build for
```
Can you open an issue about this?  I'm not sure from this comment how to articulate the problem, and I don't want to lose it

matthijskooijman replied:
Done: https://github.com/Arduino-CI/arduino_ci/issues/204

ianfixes commented:
This is in good shape to merge, and to solve a bunch of problems at once I picked it into my #183 branch
